### PR TITLE
DRAFT: Potencial fix for VideoTexture

### PR DESCRIPTION
**Pull request by eXponenta**
Related issue: #24757 

**Description**

Check texture reference in `update` method safly, null supported.

DRAFT, this is as idea!

### CODE REVIEW

File: src/textures/VideoTexture.js [hunk 1]
```diff
@@ -40,8 +27,35 @@ class VideoTexture extends Texture {
...
+
+			this._callbackSource = null;
+
+			return;
+
+		}
+
 		const hasVideoFrameCallback = 'requestVideoFrameCallback' in video;
 
+		if ( hasVideoFrameCallback && video !== this._callbackSource ) {
```
Comment (LeviPesin) [inline src/textures/VideoTexture.js:41]: You should stop the previous callback loop when creating a new one.

File: src/textures/VideoTexture.js [hunk 2]
```diff
@@ -40,8 +27,35 @@ class VideoTexture extends Texture {
 	update() {
 
 		const video = this.image;
+
+		if ( video == null ) {
```
[the remark this answers is not in the document]
Comment (eXponenta) [reply src/textures/VideoTexture.js:31]: This is not typo, source can be `undefined`. === checks stricly on null, == on `null` and `undefined`
This is special

[the remark this answers is not in the document]
Comment (LeviPesin) [reply src/textures/VideoTexture.js:31]: Then rewrite it as `( video === null ) || ( video === undefined )`.

[the remark this answers is not in the document]
Comment (eXponenta) [reply src/textures/VideoTexture.js:31]: if lint not block this, == is valid.

File: src/textures/VideoTexture.js (hunk 1 above)
Comment (eXponenta) [reply src/textures/VideoTexture.js:41]: Yep, this is problem.

We should store and callbackId and source reference, that add already 2 extra field to instance, what is undesirable

File: src/textures/VideoTexture.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (LeviPesin) [reply src/textures/VideoTexture.js:31]: Lint rules of this project do block the use of `==`.

File: src/textures/VideoTexture.js (hunk 1 above)
Comment (LeviPesin) [reply src/textures/VideoTexture.js:41]: Another way to solve would be just to check in `update` function whether the callback source and the video are still the same, and request callback only if they still are.

Comment (eXponenta) [reply src/textures/VideoTexture.js:50]: This is bad. 
requirestFrameCallback is desynced. Some video has 30 fps, and will update one of 2 of frames.
Some require tick 90+ FPS, and there are a monitors that tick 240 FPS.

update is ticks with render FPS, and you can't requiest video callback from update each call.

Comment (LeviPesin) [reply src/textures/VideoTexture.js:50]: The check can be made in the start of `update`, so if the video changes the old callback would return before setting `needsUpdate` to `true`.

> update is ticks with render FPS, and you can't requiest video callback from update each call.

But you do currently?

File: src/textures/VideoTexture.js [hunk 3]
```diff
@@ -40,8 +28,49 @@ class VideoTexture extends Texture {
...
+			if ( this._callbackId !== null && this._callbackSource !== null ) {
+
+				this._callbackSource.cancelVideoFrameCallback( this._callbackId );
+
+			}
+
+			this._callbackSource = video;
+
+			const update = () => {
+
```
Comment (LeviPesin) [inline src/textures/VideoTexture.js:61]:
```suggestion

				if ( this._callbackSource !== video ) return;

```
A simpler alternative to https://github.com/mrdoob/three.js/pull/24758/commits/14697569da22238762f69a8e7222af0f12e74d45.

**Comment by Mugen87** [comment]
Sorry, but I think the PR introduces a complexity in `VideoTexture` which is no acceptable to address #24757. Right now, I think it's best to update the documentation and state that it's not possible to change the video after the creation of a `VideoTexture`.

**Pull request closed without merging**